### PR TITLE
refactor(crypto): use pure rust symmetric crypto

### PR DESCRIPTION
**Pull request by kurnevsky**
Later xsalsa20Poly1305 can be replaced by chacha20poly1305 in most of
the cases.

### CODE REVIEW

File: tox_core/src/dht/server/mod.rs [hunk 1]
```diff
@@ -1373,9 +1376,9 @@ mod tests {
 
     use std::net::SocketAddr;
 
-    const ONION_RETURN_1_PAYLOAD_SIZE: usize = ONION_RETURN_1_SIZE - secretbox::NONCEBYTES;
-    const ONION_RETURN_2_PAYLOAD_SIZE: usize = ONION_RETURN_2_SIZE - secretbox::NONCEBYTES;
-    const ONION_RETURN_3_PAYLOAD_SIZE: usize = ONION_RETURN_3_SIZE - secretbox::NONCEBYTES;
+    const ONION_RETURN_1_PAYLOAD_SIZE: usize = ONION_RETURN_1_SIZE - xsalsa20poly1305::NONCE_SIZE;
```
Comment (kpp) [inline tox_core/src/dht/server/mod.rs:1379]: We got these consts in tox_core/src/dht/codec.rs. Why did we duplicate them?

File: tox_core/src/dht/server/mod.rs [hunk 2]
```diff
@@ -2931,19 +2934,6 @@ mod tests {
         assert!(rx.collect::<Vec<_>>().await.is_empty());
     }
 
-    #[tokio::test]
-    async fn refresh_onion_key() {
```
Comment (kpp) [inline tox_core/src/dht/server/mod.rs:2935]: Why did you remove this test?

File: tox_encryptsave/Cargo.toml [hunk 3]
```diff
@@ -21,4 +21,8 @@ edition = "2018"
 [dependencies]
 tox_crypto = { version = "0.1.0", path = "../tox_crypto" }
```
Comment (kpp) [inline tox_encryptsave/Cargo.toml:22]: no need to keep tox_crypto as dep anymore?

File: tox_encryptsave/src/lib.rs [hunk 4]
```diff
@@ -112,32 +104,30 @@ impl PassKey {
...
-                OpsLimit(ops * 2),
-                MEMLIMIT_INTERACTIVE
-            ).or(Err(KeyDerivationError::Failed))?
-        });
-
-        let salt = Box::new(salt);
-        let key = Box::new(maybe_key.ok_or(KeyDerivationError::Failed)?);
-
-        Ok(PassKey { salt, key })
+        let mut key = [0; xsalsa20poly1305::KEY_SIZE];
```
Comment (kpp) [inline tox_encryptsave/src/lib.rs:114]: Wait a minute! Can we decrypt old profile with the new code?

File: tox_core/src/dht/server/mod.rs (hunk 2 above)
Comment (kurnevsky) [reply tox_core/src/dht/server/mod.rs:2935]: Because the key is private in `XSalsa20Poly1305` struct, so we can't compare it anymore.

File: tox_core/src/dht/server/mod.rs (hunk 1 above)
Comment (kurnevsky) [reply tox_core/src/dht/server/mod.rs:1379]: They are defined in tests only, and private. Can be shared, but then need to find ath appropriate place for them.

File: tox_encryptsave/src/lib.rs (hunk 4 above)
Comment (kurnevsky) [reply tox_encryptsave/src/lib.rs:114]: Yes :)

File: tox_encryptsave/src/lib.rs [hunk 5]
```diff
@@ -423,21 +406,17 @@ fn pass_key_new_test() {
     let passwd = [42; 123];
     let pk = PassKey::from_passphrase(&passwd).expect("Failed to unwrap PassKey!");
 
-    assert_ne!(pk.salt.0.as_ref(), &passwd as &[u8]);
-    assert_ne!(pk.salt.0.as_ref(), [0; SALT_LENGTH].as_ref());
-    assert_ne!(pk.key.0.as_ref(), &passwd as &[u8]);
-    assert_ne!(pk.key.0, [0; KEY_LENGTH]);
```
Comment (kpp) [inline tox_encryptsave/src/lib.rs:429]: How about pk.key tests?

Comment (kurnevsky) [reply tox_encryptsave/src/lib.rs:429]: Same as with refreshing onion key tests - we can't access key from XSalsa20Poly1305 struct.

File: tox_encryptsave/src/lib.rs (hunk 4 above)
Comment (kpp) [reply tox_encryptsave/src/lib.rs:114]: Do we get the same results as before?

Comment (kurnevsky) [reply tox_encryptsave/src/lib.rs:114]: Yes, we have a test for this.

File: tox_encryptsave/Cargo.toml (hunk 3 above)
Comment (kurnevsky) [reply tox_encryptsave/Cargo.toml]: Yes, fixed.

File: tox_crypto/src/lib.rs [hunk 6]
```diff
@@ -4,9 +4,6 @@
 
 pub use sodiumoxide::randombytes::randombytes_into;
```
Comment (kpp) [inline tox_crypto/src/lib.rs:5]: What about randombytes_into? Why isn't it replaced with rand?

Comment (kurnevsky) [reply tox_crypto/src/lib.rs:5]: Because I missed it :)

File: tox_core/src/onion/onion_announce.rs [hunk 7]
```diff
@@ -703,7 +703,7 @@ mod tests {
             sendback_data: 42
         };
         let onion_return = OnionReturn {
-            nonce: secretbox::gen_nonce(),
+            nonce: [0; xsalsa20poly1305::NONCE_SIZE],
```
Comment (kpp) [inline tox_core/src/onion/onion_announce.rs:706]:
```suggestion
            nonce: [42; xsalsa20poly1305::NONCE_SIZE],
```

Other tests use 42 as a pseudorandom value

Comment (kurnevsky) [reply tox_core/src/onion/onion_announce.rs:706]: What do you mean?

File: tox_encryptsave/src/lib.rs [hunk 8]
```diff
@@ -109,35 +101,32 @@ impl PassKey {
...
+            &scrypt::Params::new(14, 8, 2).unwrap(),
+            &mut key[..],
+        ).or(Err(KeyDerivationError::Failed))?;
+
+        let pass_key = PassKey {
+            salt,
+            key: Box::new(XSalsa20Poly1305::new((&key).into())),
+        };
+
+        key.zeroize();
```
Comment (kpp) [inline tox_encryptsave/src/lib.rs:127]: Won't work if we get error on `scrypt().or(..)?`

File: tox_packet/src/onion/mod.rs [hunk 9]
```diff
@@ -228,13 +219,15 @@ impl ToBytes for AnnounceStatus {
...
+    use xsalsa20poly1305::aead::NewAead;
+
+    const ONION_RETURN_1_PAYLOAD_SIZE: usize = ONION_RETURN_1_SIZE - xsalsa20poly1305::NONCE_SIZE;
 
     encode_decode_test!(
         tox_crypto::crypto_init().unwrap(),
         onion_return_encode_decode,
         OnionReturn {
-            nonce: secretbox::gen_nonce(),
+            nonce: [42; 24],
```
Comment (kpp) [inline tox_packet/src/onion/mod.rs:230]: Why 24?

File: tox_encryptsave/src/lib.rs (hunk 8 above)
Comment (kurnevsky) [reply tox_encryptsave/src/lib.rs]: Right, fixed.

File: tox_packet/src/onion/mod.rs (hunk 9 above)
Comment (kurnevsky) [reply tox_packet/src/onion/mod.rs]: Changed to xsalsa20poly1305::NONCE_SIZE

File: tox_packet/src/dht/cookie.rs [hunk 10]
```diff
@@ -278,7 +283,7 @@ mod tests {
             payload: vec![43; 88]
         };
         let cookie_2 = EncryptedCookie {
-            nonce: secretbox::gen_nonce(),
+            nonce: [43; xsalsa20poly1305::NONCE_SIZE],
```
Comment (kpp) [inline tox_packet/src/dht/cookie.rs:286]: 43?)))

Comment (kurnevsky) [reply tox_packet/src/dht/cookie.rs:286]: Yes, because cookie_2 is different from cookie_1 :)

File: tox_encryptsave/src/lib.rs [hunk 11]
```diff
@@ -354,7 +335,6 @@ pub enum KeyDerivationError {
     #[fail(display = "Provided passphrase is empty")]
```
Comment (kpp) [inline tox_encryptsave/src/lib.rs:335]: Do we still get all errors of KeyDerivationError?

File: tox_encryptsave/src/lib.rs [hunk 12]
```diff
@@ -109,35 +101,30 @@ impl PassKey {
...
-        let salt = Box::new(salt);
-        let key = Box::new(maybe_key.ok_or(KeyDerivationError::Failed)?);
-
-        Ok(PassKey { salt, key })
+        let mut key = Zeroizing::new([0; xsalsa20poly1305::KEY_SIZE]);
+
+        scrypt::scrypt(
+            &passhash,
+            &salt,
+            &scrypt::Params::new(14, 8, 2).unwrap(),
```
Comment (kpp) [inline tox_encryptsave/src/lib.rs:118]: what is 14, 8, 2?

File: tox_encryptsave/src/lib.rs (hunk 11 above)
Comment (kurnevsky) [reply tox_encryptsave/src/lib.rs:335]: Yes, we have both of them.

File: tox_encryptsave/src/lib.rs (hunk 12 above)
Comment (kurnevsky) [reply tox_encryptsave/src/lib.rs:118]: Params to scrypt used by c-toxcore for key derivation.

File: tox_core/src/onion/onion_announce.rs (hunk 7 above)
Comment (kpp) [reply tox_core/src/onion/onion_announce.rs:706]: I mean `payload: vec![42; ONION_RETURN_3_PAYLOAD_SIZE]` ~ pseudorandom payload of 42*ONION_RETURN_3_PAYLOAD_SIZE. Whereas `nonce: [0; xsalsa20poly1305::NONCE_SIZE]` ~ empty value

Comment (kpp) [reply tox_core/src/onion/onion_announce.rs:706]: It does not matter, actually. It's up to you

File: tox_encryptsave/src/lib.rs (hunk 12 above)
Comment (kurnevsky) [reply tox_encryptsave/src/lib.rs:118]: They are defined in different form by Ops and Mem limits: https://docs.rs/sodiumoxide/0.2.6/sodiumoxide/crypto/pwhash/scryptsalsa208sha256/fn.derive_key.html

**Review by kpp — APPROVED** [review]

File: tox_encryptsave/src/lib.rs (hunk 12 above)
Comment (kpp) [reply tox_encryptsave/src/lib.rs:118]: Can you create consts/vars with doc?

File: tox_core/src/onion/onion_announce.rs (hunk 7 above)
Comment (kurnevsky) [reply tox_core/src/onion/onion_announce.rs]: Got it. Changed for consistency.
